sympy/solvers/ode.py

```diff
@@ -1499,6 +1481,54 @@ def _test_term(coeff, order):
     else:
         return tuple(retlist)
 
+def match_2nd_linear_bessel(r, func):
+    # eq = a3*f(x).diff(x, 2) + b3*f(x).diff(x) + c3*f(x)
+    f = func
+    x = func.args[0]
+    df = f.diff(x)
+    a = Wild('a', exclude=[f,df])
+    b = Wild('b', exclude=[x, f,df])
+    a4 = Wild('a4', exclude=[x,f,df])
+    b4 = Wild('b4', exclude=[x,f,df])
+    c4 = Wild('c4', exclude=[x,f,df])
+    d4 = Wild('d4', exclude=[x,f,df])
+    a3 = Wild('a3', exclude=[f, df, f.diff(x, 2)])
+    b3 = Wild('b3', exclude=[f, df, f.diff(x, 2)])
+    c3 = Wild('c3', exclude=[f, df, f.diff(x, 2)])
+    coeff = r[a3].match(a*(x-a4)**b4) # leading coeff of f(x).diff*x, 2)
+    # making a3 in the form of x**2
+    r[a3] = cancel(r[a3]/(coeff[a]*(x-coeff[a4])**(-2+coeff[b4])))
+    r[b3] = cancel(r[b3]/(coeff[a]*(x-coeff[a4])**(-2+coeff[b4])))
+    r[c3] = cancel(r[c3]/(coeff[a]*(x-coeff[a4])**(-2+coeff[b4])))
+    # checking if b3 is of fome c*(x-b)
+    coeff1 = r[b3].match(a4*(x-b4))
+    if coeff1 is None:
+        return None
+    # c3 maybe of very complex form so I am simply checking (a - b) form
+    # if yes later I will match with the standerd form of bessel in a and b
+    # a, b are wild variable defined above.
+    _coeff2 = r[c3].match(a - b)
+
+    if _coeff2 is None:
+        return None
+    # matching with standerd form for c3
+    coeff2 = _coeff2[a].match(c4**2*(x-b4)**(2*a4))
+    if coeff2 is None:
+        return None
+
+    if _coeff2[b] == 0:
+        coeff2[d4] = 0
+    else:
+         coeff2[d4] = _coeff2[b].match(d4**2)[d4]
+
+    if (coeff1[b4] != coeff2[b4]) or (coeff1[b4] != coeff[a4]):
+        return None
+
+    rn = {'n':coeff2[d4], 'a4':coeff2[c4], 'd4':coeff2[a4]}
+    rn['c4'] = coeff1[a4]
+    rn['b4'] = coeff1[b4]
+    return rn
+
```

Comment:
I have made this function because `match `is not working of the complex equations so I am breaking equations into parts and the try to match the equation.
Still, It is not working for the equation of the type -
```
In [23]: eq = (x-2)**2*(f(x).diff(x, 2)) + (x-2)*(f(x).diff(x)) + ((x-2)**2 - 4)*f
    ...: (x)

In [24]: eq
Out[24]: 
           2                                               
       2  d                  d          ⎛       2    ⎞     
(x - 2) ⋅───(f(x)) + (x - 2)⋅──(f(x)) + ⎝(x - 2)  - 4⎠⋅f(x)
           2                 dx                            
         dx                                                


```

And the reason is `match ` is not able to match.

Reply:
Maybe it's worth trying to fix match rather than work around its limitations

Reply:
Maybe but till now no-one had tried. So I can't wait.
And also it is tough to match the given equation in the required form. Because many variations of a given equation are possible. For this, we need to make a standard form of the equation, but by doing this, many specific types of matching will fail. I have seen that many times just by doing a few variations in the equation matching fails. Also, We are doing expand(eq) in the starting, and this line makes the given equation very complicated.
(See the above example)If we expand this, then I think it is very challenging to match in the required form.

Reply:
> Maybe but till now no-one had tried. So I can't wait.

You could take a look at the matching code yourself. Maybe it's not that hard to fix. There are a few different matching functions around. None of them looks especially complicated:
```console
$ git grep 'def matches'
sympy/core/add.py:    def matches(self, expr, repl_dict={}, old=False):
sympy/core/basic.py:    def matches(self, expr, repl_dict={}, old=False):
sympy/core/basic.py:    def matches(self, expr, repl_dict={}, old=False):
sympy/core/function.py:    def matches(self, expr, repl_dict={}, old=False):
sympy/core/mul.py:    def matches(self, expr, repl_dict={}, old=False):
sympy/core/power.py:    def matches(self, expr, repl_dict={}, old=False):
sympy/core/symbol.py:    def matches(self, expr, repl_dict={}, old=False):
sympy/utilities/runtests.py:    def matches(self, x):
```

Reply:
okay, I will try